critwall/src/main/java/org/opencadc/critwall/FileSyncJob.java

```diff
@@ -0,0 +1,258 @@
+        }
+    }
+
+    private List<URL> getdownloadURLs()
+        throws IOException, InterruptedException,
+        ResourceAlreadyExistsException, ResourceNotFoundException,
+        TransientException, TransferParsingException {
+
+        RegistryClient regClient = new RegistryClient();
+        log.debug("resource id: " + this.resourceID);
+        URL transferURL = regClient.getServiceURL(this.resourceID, Standards.VOSPACE_SYNC_21, AuthMethod.CERT);
+        log.debug("certURL: " + transferURL);
+
+        // Ask for all protocols available back, and the server will
+        // give you URLs for the ones it allows.
+        List<Protocol> protocolList = new ArrayList<Protocol>();
+        Protocol httpsCert = new Protocol(VOS.PROTOCOL_HTTPS_GET);
+
+        log.debug("protocols: " + httpsCert);
+        protocolList.add(httpsCert);
+
+        Transfer transfer = new Transfer(artifactID, Direction.pullFromVoSpace, protocolList);
+        transfer.version = VOS.VOSPACE_21;
+
+        TransferWriter writer = new TransferWriter();
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        writer.write(transfer, out);
+        FileContent content = new FileContent(out.toByteArray(), "text/xml");
+        log.debug("xml file content to be posted: " + transfer);
+
+        log.debug("artifact path: " + artifactID.getPath());
+        HttpPost post = new HttpPost(transferURL, content, true);
+        post.prepare();
+        log.debug("post prepare done");
+
+        TransferReader reader = new TransferReader();
+        Transfer t = reader.read(post.getInputStream(), null);
+        List<String> urlStrList = t.getAllEndpoints();
+        log.debug("endpoints returned: " + urlStrList);
+
+        List<URL> urlList = new ArrayList<URL>();
+
+        // Create URL list to return
+        for (int i = 0; i < urlStrList.size(); i++) {
+            try {
+                urlList.add(new URL(urlStrList.get(i)));
+            } catch (MalformedURLException mue) {
+                log.info("Malformed URL returned from transfer negotiation: " + urlStrList.get(i) + " skipping... ");
+            }
+        }
+
+        return urlList;
+    }
+
+    private StorageMetadata syncArtifact(Iterator<URL> urlIterator)
+        throws StorageEngageException, InterruptedException {
+
+        StorageMetadata storageMeta = null;
+
+        while (urlIterator.hasNext()) {
+            URL u = urlIterator.next();
+            log.debug("trying " + u);
+
+            try {
+                ByteArrayOutputStream dest = new ByteArrayOutputStream();
+                HttpGet get = new HttpGet(u, dest);
+                get.prepare();
+
+                NewArtifact a = new NewArtifact(artifactID);
+                storageMeta = storageAdapter.put(a, get.getInputStream());
+                log.debug("storage meta returned: " + storageMeta.getStorageLocation());
+                return storageMeta;
+
+            } catch (WriteException wre) {
```

Comment:
I think we have to consider WriteException fatal so it needs to be thrown (retry isn't likely to help in the short term). You can just throw it from the catch and subsequent catches don't interfere.

So returning null means that all URLs were tried without success but without a fatal error. I guess that works but please add a comment for the method like:
```
// success: return StorageMetadata
// fail: return null if all URLs failed | throw if failure is fatal (internal)

Reply:
WriteException thrown, caught in outer try/catch for error log before end of run().

comment added for the method.